### PR TITLE
CSPDARKNET Subclassing 

### PR DESCRIPTION
**Pull request by soma2000-lang**
[colab](https://colab.research.google.com/drive/12XKas34I6eBmAMBjckFzwLDtQubnxskG?usp=sharing)
#1420
@ianstenbit

### CODE REVIEW

File: keras_cv/models/csp_darknet.py [hunk 1]
```diff
@@ -36,22 +36,9 @@
...
-    input_tensor=None,
-    pooling=None,
-    classifier_activation="softmax",
-    name="CSPDarkNet",
-    **kwargs,
-):
-    """Instantiates the CSPDarkNet architecture.
+@keras.utils.regester_keras_serializable(package="keras_cv.models")
+class CSPDarkNet(keras.Model):
+    """This class Instantiates the CSPDarkNet architecture.
```
Comment (ianstenbit) [inline keras_cv/models/csp_darknet.py:41]: nit: s/Instantiates/represents

File: keras_cv/models/csp_darknet.py [hunk 2]
```diff
@@ -266,8 +276,8 @@ def CSPDarkNet(
 
 def CSPDarkNetTiny(
     *,
-    include_rescaling,
-    include_top,
+    include_rescaling=None,
```
Comment (ianstenbit) [inline keras_cv/models/csp_darknet.py:279]: let's avoid including defaults here

**Review by ianstenbit — CHANGES_REQUESTED** [review]
Thanks for the PR!

Let's also remove the closures from `darknet_utils`, and use the `apply_xyz` naming technique (this will also need to be done in #1486)

**Comment by ID6109** [comment]
Hey @soma2000-lang , this is in reference to #1486 . I have already started working on `darknet_utils.py`, so how about I get back to you once I am through with the changes that can be incorporated into your PR? Thanks.

**Comment by soma2000-lang** [comment]
Ok @ID6109 seems good

**Comment by ID6109** [comment]
@soma2000-lang I have made the necessary changes in `darknet_utils.py` but the tests are failing due to function name mismatches with `csp_darknet.py`. How about I go ahead with making the naming changes there as well? This might cause merge conflicts with your PR.

**Comment by soma2000-lang** [comment]
@ID6109 let me make the name changes.
So that it does not cause the merge conflicts.Please tell me what name changes do you want to do

**Comment by soma2000-lang** [comment]
@ID6109 I will do that

**Comment by ianstenbit** [comment]
> Let's also remove the closures from `darknet_utils`, and use the `apply_xyz` naming technique (this will also need to be done in #1486)

Upon further inspection, let's _not_ update the layers in `darknet_utils` and instead leave them as-is.

So this PR is very close -- just the few small comments I left.

Additionally, let's test pre-trained weights in the Colab. `"imagenet"` weights are available for `CSPDarkNetL`

File: keras_cv/models/csp_darknet.py [hunk 3]
```diff
@@ -267,7 +277,7 @@ def CSPDarkNet(
 def CSPDarkNetTiny(
     *,
     include_rescaling,
-    include_top,
+    include_top=None,
```
Comment (ianstenbit) [inline keras_cv/models/csp_darknet.py:280]: nit: please remove this default value of `include_top`

**Review by ianstenbit — APPROVED** [review]
Looks good -- just one more small fix and then we can merge :)

Your Colab no longer runs, but I was able to adapt it to make sure weight loading still works:
https://colab.research.google.com/gist/ianstenbit/b2ef30c13329624e45200008ce4acb46/cspdarknetm.ipynb

With respect to KerasIO guides -- right now, we're still a bit short on model offerings and we're still in the process of updating our API to unify with KerasNLP, so we're holding off on guides for`KerasCV` models. However, if there are models that we don't include in KerasCV (yet!), feel free to create guides on those, and we can use that as a starting point for introducing those models here.

**Comment by ianstenbit** [comment]
Also @soma2000-lang please take a look at the lint and test errors

File: keras_cv/models/csp_darknet.py [hunk 4]
```diff
@@ -36,22 +36,9 @@
...
-    weights=None,
-    input_shape=(None, None, 3),
-    input_tensor=None,
-    pooling=None,
-    classifier_activation="softmax",
-    name="CSPDarkNet",
-    **kwargs,
-):
-    """Instantiates the CSPDarkNet architecture.
+@keras.utils.regester_keras_serializable(package="keras_cv.models")
```
Comment (ianstenbit) [inline keras_cv/models/csp_darknet.py:39]: nit: s/regester/register

File: keras_cv/models/csp_darknet.py [hunk 5]
```diff
@@ -257,12 +136,134 @@ def CSPDarkNet(
...
+            x = layers.GlobalMaxPooling2D(name="max_pool")(x)
+
+        model = keras.Model(inputs, x, name=name, **kwargs)
+        model._backbone_level_outputs = _backbone_level_outputs
+        # Bind the `to_backbone_model` method to the application model.
+        model.as_backbone = types.MethodType(utils.as_backbone, model)
+
+        if weights is not None:
+            model.load_weights(weights)
+        return model
```
Comment (ianstenbit) [inline keras_cv/models/csp_darknet.py:265]: Sorry I missed this in the first review.

We shouldn't be `return`ing anything, as this is now a constructor.

We do need to set model-level properties, though, and add a `get_config` and `from_config`

See e.g. https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/regnet.py#L791 for an example of this

**Review by ianstenbit — CHANGES_REQUESTED** [review]

**Pull request merged**